dashboard/app/models/pd/attendance.rb

```diff
@@ -57,13 +58,18 @@ def self.distinct_teachers
   end
 
   # Idempotent: Find existing, restore deleted, or create a new attendance row.
-  def self.find_restore_or_create_by!(attendance_params)
+  # @param search_params [Hash] params to search, or create by
+  # @param then_update [Hash] optional additional params to set after retrieving or creating from the search params.
+  # @return [Pd::Attendance] resulting attendance model
+  def self.find_restore_or_create_by!(search_params, then_update: {})
     attendance = nil
     Retryable.retryable(on: ActiveRecord::RecordNotUnique) do
-      attendance = Pd::Attendance.with_deleted.find_by(attendance_params) ||
-        Pd::Attendance.create!(attendance_params)
+      attendance = Pd::Attendance.with_deleted.find_by(search_params) ||
+        Pd::Attendance.new(search_params)
     end
-    attendance.restore if attendance.deleted?
+
+    attendance.assign_attributes(then_update.merge(deleted_at: nil))
+    attendance.save! if attendance.changed?
```

Comment:
The main reason was I was attempting to reduce unnecessary DB calls. Here it restores (if deleted) and only makes the `then_update` changes if they are actually changes from the DB state. Otherwise it's a no-op. It's probably not that important, so I could pull out the update, but then note it will always be a DB call, in addition to potentially the `restore` call.